packages/devtools-evm-hardhat/src/tasks/healthcheck/validate-rpcs.ts

```diff
@@ -0,0 +1,161 @@
+    return new Promise((resolve) => {
+        const ws = new WebSocket(rpcUrl)
+        const timeoutId = setTimeout(() => {
+            ws.close()
+            logger.error(`WebSocket connection timed out for RPC URL`)
+            resolve(false)
+        }, timeout)
+
+        const cleanup = (success: boolean, message?: string) => {
+            clearTimeout(timeoutId)
+            if (message) {
+                logger.error(message)
+            }
+            ws.close()
+            resolve(success)
+        }
+
+        ws.onopen = () => {
+            const rpcRequest = JSON.stringify({
+                jsonrpc: JSON_RPC,
+                method: 'eth_blockNumber',
+                params: [],
+                id: 1,
+            })
+            ws.send(rpcRequest)
+        }
+
+        ws.onmessage = (event) => {
+            try {
+                const response = JSON.parse(event.data.toString())
+                if (response.jsonrpc === JSON_RPC && !!response.result) {
+                    cleanup(true)
+                } else {
+                    cleanup(false, `Invalid RPC response: ${event.data}`)
+                }
+            } catch (error) {
+                cleanup(false, `Error parsing RPC response: ${error}`)
+            }
+        }
+
+        ws.onerror = (error) => {
+            cleanup(false, `Validation failed: ${error instanceof Error ? error.message : 'An unknown error occurred'}`)
+        }
+
+        ws.onclose = () => {
+            cleanup(false, `Validation failed: WebSocket connection closed unexpectedly`)
+        }
+    })
+}
+
+const validateRpcUrl = async (rpcUrl: string | undefined, timeout: number, networkName: string): Promise<boolean> => {
+    if (!rpcUrl || rpcUrl.trim() === '') {
+        logger.error(`Missing RPC URL for network: ${networkName}`)
+        return false
+    }
+
+    logger.info(`...Validating RPC URL ${rpcUrl} for ${networkName}...`)
+
+    if (rpcUrl.startsWith(HTTP_URL) || rpcUrl.startsWith(HTTPS_URL)) {
+        return await validateHttpsRpcUrl(rpcUrl, timeout)
+    } else if (rpcUrl.startsWith(WS_URL) || rpcUrl.startsWith(WSS_URL)) {
+        return await validateWebSocketRpcUrl(rpcUrl, timeout)
+    }
+
+    logger.error(`Unsupported RPC protocol in network: ${networkName}`)
+
+    return false
+}
+
+const action: ActionType<TaskArguments> = async (taskArgs, hre) => {
+    printLogo()
+
+    const networks = hre.userConfig.networks || {}
+    const networkNames = Object.keys(networks)
+
+    logger.info(
+        `========== Validating RPC URLs with ${taskArgs.timeout}ms timeout for networks: ${networkNames.join(', ')}`
+    )
+
+    for (const networkName of networkNames) {
+        const rpcUrl = networks[networkName]?.[RPC_URL_KEY]
+
+        if (rpcUrl && !(await validateRpcUrl(rpcUrl, taskArgs.timeout, networkName))) {
+            throw new Error(`${printBoolean(false)} RPC URL validation failed for network: ${networkName}`)
+        }
+    }
```

Comment:
This will validate the URLs one by one - it might be worth using `Promise.all` instead to parallelize

Reply:
Ah I see what you mean -- then the user will know all the bad URLs in one go instead of needing to re-run the task after each fix. I will look into this!